### DISCUSSION TITLE
Check if ghidra process is already running

### DISCUSSION THREAD

**Discussion by chiffreKing** [NationalSecurityAgency/ghidra#3640] [Q&A]
Hi guys,

I'm writing a program that is using ghidra. Before I start ghidra + ghidra script as a subprocess I want to check if a non headless ghidra instance of the same project is already running to avoid that the headless mode will fail due to two running ghidra projects at the same time. 
I figuered out that the running process of ghidra (using Windows 10) is javaw.exe . The problem is there are multiple javaw.exe 's running and only one of them is ghidra so I don't know which one to close. Is there any other way to figure out the opened ghidra process? Thanks.

**Comment by ryanmkurtz** [comment]
How are you examining the processes? You should be able to get the command line arguments that were given to each process at startup. If you can get that, you should be able to easily tell which one is Ghidra.

**Reply by chiffreKing** [reply]
Alright thanks, this helped me out.

My solution is to determine the process with 'psutil' and then use the cmdline() methode and filter the resulting list for 'ghidra.GhidraRun'. if found then i get the processID and kill the process.

**Reply by astrelsky** [reply]
> Alright thanks, this helped me out.
> 
> My solution is to determine the process with 'psutil' and then use the cmdline() methode and filter the resulting list for 'ghidra.GhidraRun'. if found then i get the processID and kill the process.

While this may work for your specific case it is possible to use a `GhidraLaunchable` other than `GhidraRun`. I recommend instead checking for the classloader argument and `GhidraClassLoader` which must always be specified.

**Reply by chiffreKing** [reply]
> 
> 
> > Alright thanks, this helped me out.
> > My solution is to determine the process with 'psutil' and then use the cmdline() methode and filter the resulting list for 'ghidra.GhidraRun'. if found then i get the processID and kill the process.
> 
> While this may work for your specific case it is possible to use a `GhidraLaunchable` other than `GhidraRun`. I recommend instead checking for the classloader argument and `GhidraClassLoader` which must always be specified.

So you mean that the java.system.class.loader always contains the arguement 'GhidraClassLoader' ?

This is what it shows:
'-Djava.system.class.loader=ghidra.GhidraClassLoader'

**Reply by astrelsky** [reply]
> > > Alright thanks, this helped me out.
> > > My solution is to determine the process with 'psutil' and then use the cmdline() methode and filter the resulting list for 'ghidra.GhidraRun'. if found then i get the processID and kill the process.
> > 
> > 
> > While this may work for your specific case it is possible to use a `GhidraLaunchable` other than `GhidraRun`. I recommend instead checking for the classloader argument and `GhidraClassLoader` which must always be specified.
> 
> So you mean that the java.system.class.loader always contains the arguement 'GhidraClassLoader' ?
> 
> This is what it shows: '-Djava.system.class.loader=ghidra.GhidraClassLoader'

Well it turns out that it just needs to be an instanceof `ghidra.GhidraClassLoader` but if you ever encounter a case where it isn't (which it won't be if your application is launching Ghidra) then you could handle it then.

https://github.com/NationalSecurityAgency/ghidra/blob/0996ede391c0264ccf1965693ccc115aadd20fc6/Ghidra/Framework/Utility/src/main/java/ghidra/GhidraLauncher.java#L91-L96
